### PR TITLE
mon: osd [test-]reweight-by-{pg,utilization} command updates

### PR DESCRIPTION
**Pull request by liewegas**
- 'osd utilization' command to show current pg distribution stats
- show before/after stats
- test-* variants
- --no-increasing option for the -utilization ones

### CODE REVIEW

**Comment by liewegas** [comment]
@dvanders

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -560,61 +559,127 @@ int OSDMonitor::reweight_by_utilization(int oload, std::string& out_str,
...
     }
-    if (util <= underload_util) {
+    if (!no_increasing && util <= underload_util) {
       // assign a higher weight.. if we can.
       unsigned weight = osdmap.get_weight(p->first);
       unsigned new_weight = (unsigned)((average_util / util) * (float)weight);
+      new_weight = MIN(new_weight, weight + max_change);
       if (new_weight > 0x10000)
-	new_weight = 0x10000;
+          new_weight = 0x10000;
```
Comment (athanatos) [inline src/mon/OSDMonitor.cc]: wrong indentation?

File: doc/man/8/ceph.rst [hunk 2]
```diff
@@ -982,6 +982,7 @@ Subcommand ``reweight-by-pg`` reweight OSDs by PG distribution
 Usage::
 
 	ceph osd reweight-by-pg {<int[100-]>} {<poolname> [<poolname...]}
+	{--no-increasing} {--yes-i-really-mean-it}
```
Comment (athanatos) [inline doc/man/8/ceph.rst]: Is the --yes-i-really-mean-it still there?

File: src/mon/MonCommands.h [hunk 3]
```diff
@@ -701,13 +701,31 @@ COMMAND("osd pool stats " \
...
+	"name=oload,type=CephInt,range=100,req=false " \
+	"name=max_change,type=CephFloat,req=false "			\
+	"name=no_increasing,type=CephChoices,strings=--no-increasing,req=false",\
+	"reweight OSDs by utilization [overload-percentage-for-consideration, default 120]", \
+	"osd", "rw", "cli,rest")
+COMMAND("osd test-reweight-by-utilization " \
+	"name=oload,type=CephInt,range=100,req=false " \
+	"name=max_change,type=CephFloat,req=false "			\
+	"name=no_increasing,type=CephChoices,strings=--no-increasing,req=false",\
 	"reweight OSDs by utilization [overload-percentage-for-consideration, default 120]", \
```
Comment (athanatos) [inline src/mon/MonCommands.h]: Different help text?

File: src/mon/MonCommands.h [hunk 4]
```diff
@@ -701,13 +701,31 @@ COMMAND("osd pool stats " \
...
+	"name=oload,type=CephInt,range=100,req=false " \
+	"name=max_change,type=CephFloat,req=false "			\
+	"name=pools,type=CephPoolname,n=N,req=false",			\
+	"reweight OSDs by PG distribution [overload-percentage-for-consideration, default 120]", \
+	"osd", "rw", "cli,rest")
+COMMAND("osd test-reweight-by-pg " \
+	"name=oload,type=CephInt,range=100,req=false " \
+	"name=max_change,type=CephFloat,req=false "			\
+	"name=pools,type=CephPoolname,n=N,req=false",			\
 	"reweight OSDs by PG distribution [overload-percentage-for-consideration, default 120]", \
```
Comment (athanatos) [inline src/mon/MonCommands.h]: Also here.

File: src/mon/OSDMonitor.cc [hunk 5]
```diff
@@ -478,16 +478,19 @@ void OSDMonitor::update_logger()
  * The osds that will get a lower weight are those with with a utilization
  * percentage 'oload' percent greater than the average utilization.
  */
-int OSDMonitor::reweight_by_utilization(int oload, std::string& out_str,
-					bool by_pg, const set<int64_t> *pools)
+int OSDMonitor::reweight_by_utilization(int oload,
+					double max_changef,
+					bool by_pg, const set<int64_t> *pools,
+					bool no_increasing, bool sure,
```
Comment (athanatos) [inline src/mon/OSDMonitor.cc]: sure isn't a great param name here.  How about dry_run?

File: src/common/config_opts.h [hunk 6]
```diff
@@ -281,6 +281,7 @@ OPTION(mon_daemon_bytes, OPT_U64, 400ul << 20)  // mds, osd message memory cap (
 OPTION(mon_max_log_entries_per_event, OPT_INT, 4096)
 OPTION(mon_reweight_min_pgs_per_osd, OPT_U64, 10)   // min pgs per osd for reweight-by-pg command
 OPTION(mon_reweight_min_bytes_per_osd, OPT_U64, 100*1024*1024)   // min bytes per osd for reweight-by-utilization command
+OPTION(mon_reweight_max_osds, OPT_INT, 4)   // max osds to change per reweight-by-* command
```
Comment (cernceph) [inline src/common/config_opts.h]: Would this be better on the CLI? --max-osds

File: src/mon/OSDMonitor.cc [hunk 7]
```diff
@@ -7449,18 +7514,36 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
       }
       pools.insert(pool);
     }
+    double max_change = .05;
```
Comment (athanatos) [inline src/mon/OSDMonitor.cc]: This should probably be a config value.

**Comment by athanatos** [comment]
@liewegas comments

**Comment by liewegas** [comment]
@athanatos updated.  I didn't squash the max_change fixes down, though, because it crosses a refactor commit and it's tedious.

**Comment by athanatos** [comment]
Fair enough, taking another look.

**Comment by athanatos** [comment]
lgtm

**Comment by liewegas** [comment]
pushed a fix, needs re-run

File: doc/man/8/ceph.rst (hunk 2 above)
Comment (tchaikov) [inline doc/man/8/ceph.rst]: @athanatos https://github.com/ceph/ceph/pull/8024

**Pull request merged**